CHANGE: Override UGUI EventSystem with InputSystemUIInputModule

jfreire-unity opened the pull request:
### Description

Unity 6 will come with the "new" Input System enabled by default. The `EventSystem` of UGUI uses `StandaloneInputModule` regardless of the enabled input backend.
This PR overrides the `EventSystem` game object default InputModule using a new class,  `InputModuleComponentFactory` , that will be available on Unity 6.

### Changes made

- Event System Input Module is overridden with `InputSystemUIInputModule` when the Input System native backend is enabled. Only with "Legacy Input"  in "Project Settings > Player > Active Input Handling", no override is done.
- Adds an Editor test to validate the behavior we want for `EventSystem`. Once the Input System package is enabled, we want the `EventSystem` to be created with `InputSystemUIInputModule`.

### Notes

PR is in a good state for reviewing the changes it does. After that, it still needs: 
- Check if there is documentation that needs to be updated.

### Checklist

Before review:

- [x] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [x] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [x] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

jfreire-unity committed:
Override InputModule component of EventSystem

jfreire-unity committed:
Make it override only when ISX is enabled and IM disabled

InputSystem should override the input module for the `EventSystem` game object only when it's the only input system available. If IM (Input Manager) is available as well or is the only input backend available, we fallback to default behavior of adding `StandaloneInputModule`.

benoitalain commented on Packages/com.unity.inputsystem/InputSystem/Plugins/UI/InputSystemUIInputModuleEditor.cs:
```diff
@@ -2,14 +2,25 @@
+        static InputSystemUIInputModuleEditor()
+        {
+#if UNITY_6000_0_OR_NEWER
+            if (EditorPlayerSettingHelpers.newSystemBackendsEnabled && !EditorPlayerSettingHelpers.oldSystemBackendsEnabled)
```
OK, yes, makes for sense this way!

jfreire-unity replied:
Changed this to check for `ENABLE_INPUT_SYSTEM` instead.

benoitalain commented on Packages/com.unity.inputsystem/InputSystem/Plugins/UI/InputSystemUIInputModuleEditor.cs:
```diff
@@ -2,14 +2,25 @@
     {
+        static InputSystemUIInputModuleEditor()
+        {
+#if UNITY_6000_0_OR_NEWER
```
Indeed! Any reason to worry about 6000.0.0b15 and below not working with this version of the InputSystem package?

jfreire-unity replied:
Good point. So far I believe this kind of pattern is what we've been doing, which means we "don't care" about patch versions that don't have these changes. I believe we always assume the latest patch version of an Editor version because that's where the latest Input System package gets pulled cc @ekcoh .
I don't think for example: for a 6b1 version we would pull the latest 1.8.2 to be released for example. Or am I wrong? cc @stefanunity @Pauliusd01

benoitalain commented on Packages/com.unity.inputsystem/InputSystem/Plugins/UI/InputSystemUIInputModuleEditor.cs:
```diff
@@ -2,14 +2,25 @@
+        {
+#if UNITY_6000_0_OR_NEWER
+            if (EditorPlayerSettingHelpers.newSystemBackendsEnabled && !EditorPlayerSettingHelpers.oldSystemBackendsEnabled)
+                InputModuleComponentFactory.SetInputModuleComponentOverride(go => go.AddComponent<InputSystemUIInputModule>());
```
I did some more testing on uGUI. This will have to be replaced by
```
ObjectFactory.AddComponent<InputSystemUIInputModule>(go)
```
, otherwise we have a weird bug with Undo. See gif below:

![Unity_9y36X5MqV9](https://github.com/Unity-Technologies/InputSystem/assets/35855598/eda37f8a-6f78-463a-991c-06fc717b6d7a)

jfreire-unity replied:
Good catch ⚡ Replaced it

jfreire-unity committed:
Use static ObjectFactory.AddComponent

jfreire-unity committed:
Test that UGUI EventSystem is created correctly

This test assumes that the Input System package is enabled in the
project. In cases like this, the EventSystem GameObject will always
be created with the InputSystemUIInputModule.

jfreire-unity committed:
Rely on ENABLE_INPUT_SYSTEM instead

This avoids us having to use `EditorPlayerSettingsHelpers` which uses `FindObjectsOfTypeAll`. This is not allowed to be done on ScriptableObject constructor (or instance field initializer)

jfreire-unity committed:
Update CHANGELOG

jfreire-unity commented:
⚠️ Note that this PR requires changes on `trunk` to land first. That's why trunk tests are currently failing.

ekcoh reviewed (approved):
Thanks for addressing this issue. Changes looks good to me.

benoitalain commented:
Tested with https://github.cds.internal.unity3d.com/unity/unity/pull/47422 changes to trunk and this works. Now we only need to wait for that PR to land.

timkeo commented on Packages/com.unity.inputsystem/InputSystem/Plugins/UI/InputSystemUIInputModuleEditor.cs:
```diff
@@ -2,14 +2,25 @@
 namespace UnityEngine.InputSystem.UI.Editor
 {
     [CustomEditor(typeof(InputSystemUIInputModule))]
+    [InitializeOnLoad]
```
Is this functionality affected by Fast Enter Play Mode, i.e. will this still work if Domain Reloads are disabled?

timkeo reviewed (approved):
LGTM

lyndon-unity committed:
Merge branch 'develop' into uum-68663-use-ugui-inputmoduledelegate
